internal/pkg/api/handleAck.go

```diff
@@ -503,13 +503,33 @@ func (ack *AckT) handleUnenroll(ctx context.Context, zlog zerolog.Logger, agent
 	return nil
 }
 
-func (ack *AckT) handleUpgrade(ctx context.Context, zlog zerolog.Logger, agent *model.Agent) error {
-
+func (ack *AckT) handleUpgrade(ctx context.Context, zlog zerolog.Logger, agent *model.Agent, event Event) error {
 	now := time.Now().UTC().Format(time.RFC3339)
 	doc := bulk.UpdateFields{
 		dl.FieldUpgradeStartedAt: nil,
+		dl.FieldUpgradeStatus:    nil,
 		dl.FieldUpgradedAt:       now,
 	}
+	if event.Error != "" {
+		// unmarshal event payload
+		var pl struct {
+			Retry   bool `json:"retry"`
+			Attempt int  `json:"retry_attempt"`
+		}
+		err := json.Unmarshal(event.Payload, &pl)
+		if err != nil {
+			zlog.Error().Err(err).Msg("unable to unmarshal upgrade event payload")
+		}
+
+		// if the payload indicates a retry, mark change the upgrade status to retrying.
+		if pl.Retry {
+			zlog.Info().Int("retry_attempt", pl.Attempt).Msg("marking agent upgrade as retrying")
+			doc[dl.FieldUpgradeStatus] = "retrying" // TODO should we also change FieldUpgradedAt and FieldUpgradeStated at?
+		} else {
+			zlog.Info().Int("retry_attempt", pl.Attempt).Msg("marking agent upgrade as failed")
```

Comment:
how will we handle stale agent (older versions without retry in ack)? is it expected for them not to retry?
i'd say yes, just want to double check with your intentions.

Reply:
`FieldUpgradeStartedAt` and `FieldUpgradedAt` is used in Fleet UI side to set status `Upgrading` 
i wonder if we could mark `FieldUpgradedAt` in case of failed retry as it was not upgraded so far.

Reply:
Yeah, a pre v8.6 agent that reports an error will set `FieldUpgradeStatus` to `failed`. However I'm not sure if it is able to send that ack as I don't think it could actually ack a failure.

If it sends a successful upgrade ack `FieldUpgradeStatus` will be set to nil and `FieldUpgradedAt` is set to the current time (keeping the current behaviour)

@juliaElastic should we be setting `FieldUpgradedAt` and clearing `FieldUpgradeStartedAt` in failure cases, or should we keep these attributes to their original values?

Reply:
Currently there is no `Failed` agent status. So the agent can either remain in `Updating` or move back to `Healthy` status. The error should be captured in action results similar to other action failures. 
So I think the agent can move back to `Healthy` state if the upgrade failed (supposing nothing changed for the agent) - this means setting `FieldUpgradeStartedAt` to nil and leaving `FieldUpgradedAt` at the previous value.